webpack/assets/javascripts/react_app/common/hooks/useSubmitOnEnter.js

```diff
@@ -0,0 +1,23 @@
+import { useEffect } from 'react';
+
+/**
+ * A custom hook that listens for enter on a ref and runs a function on enter keydown
+ * @param  {object} ref a ref created with React's useRef or createRef that scopes the event listener
+ * @param  {function} onSubmit function to execute on enter keydown
+ */
+export const useSubmitOnEnter = (ref, onSubmit) => {
+  const listener = event => {
+    if (event.code === 'Enter' || event.code === 'NumpadEnter') onSubmit();
+  };
+
+  useEffect(() => {
+    const refPointer = ref; // re-assigning so it's available on unmount
+    /* eslint-disable-next-line no-unused-expressions */
+    refPointer?.current?.addEventListener?.('keydown', listener);
+
+    return () =>
+      refPointer?.current?.removeEventListener?.('keydown', listener);
+  }, []);
```

Comment:
This should depend on `listener` and `ref` right? or is there a reason for you not to depend on those? :thinking:

Reply:
It just makes sense to do it on mount and clean on unmount, otherwise we are just re-attaching the same event listener

Reply:
Well that would be what happens, right? IIUR, the only difference would be in case we change the listener or the ref (we rerender) we would remove old and add new listener.

Reply:
@ezr-ondrej I was running into attaching multiple event listeners so I made it only attach on mount (this was also suggested earlier in the review)

I found event listeners inside of `useEffect` or a callback function are a bit tricky. If you pass in the exact same args to addEventListener, the browser [ignores it](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#Multiple_identical_event_listeners), so you think you can call `addEventListener` again and it would be a noop. BUT if we are defining a function inside of a callback, that will be a new function reference that is passed, so now you have another event listener attached that does the same thing because of the "new" function.

I tried a couple different iterations and eventually landed on doing it on mount and cleaning up on dismount. We want an event listener to be attached, cleaned up after, and only attached once, so I felt it fits the goal.

As an example, this will attach an event listeners on each keystroke (I also tried moving listener outside of useEffect)
```diff
 export const useSubmitOnEnter = (ref, onSubmit) => {
-  const listener = event => {
-    if (event.code === 'Enter' || event.code === 'NumpadEnter') onSubmit();
-  };
-
   useEffect(() => {
+    const listener = event => {
+      if (event.code === 'Enter' || event.code === 'NumpadEnter') onSubmit();
+    };
+
     const refPointer = ref; // re-assigning so it's available on unmount
     /* eslint-disable-next-line no-unused-expressions */
     refPointer?.current?.addEventListener?.('keydown', listener);
 
     return () =>
       refPointer?.current?.removeEventListener?.('keydown', listener);
-  }, []);
+  }, [ref]);
 };
```

I think we should stay with only attaching on mount, but open to other suggestions that account for this multiple event listener issue.

Reply:
I feel like `useEffect` requires a new way of thinking about effects.

I found [this comment](https://github.com/facebook/react/issues/14920#issuecomment-470913287) and the related discussions from React folks helpful.  (also [this one](https://github.com/facebook/react/issues/14920#issuecomment-470945051) from @gaeron himself)

Old way:  How do I make `useEffect` behave like `componentDidMount?`  Put `[]` in the dependency array.

New way: This function _is going to run_ on every render, but I only need to make changes when _X_ is the case.  What tools do I have to ensure that?

1. Yes, we have the dependency array, and we should put everything in there that's supposed to be there.  (`listener` and `ref` in our case)
2. But we also can do some old-fashioned condition checking. something like:

```js
if (ref && !listenerAdded) {
  refPointer?.current?.addEventListener?.('keydown', listener);
}
```

This will allow the effect to run on every render, but only make changes when needed.

Reply:
Expanding on this (mostly because I found it interesting), the hook is called multiple times, seems like on each render, but the useEffect is only called based on the dependency array

For example, this change:
```diff
diff --git a/webpack/assets/javascripts/react_app/common/hooks/useSubmitOnEnter.js b/webpack/assets/javascripts/react_app/common/hooks/useSubmitOnEnter.js
index 2a57ad4..f152c7c 100644
--- a/webpack/assets/javascripts/react_app/common/hooks/useSubmitOnEnter.js
+++ b/webpack/assets/javascripts/react_app/common/hooks/useSubmitOnEnter.js
@@ -10,7 +10,11 @@ export const useSubmitOnEnter = (ref, onSubmit) => {
     if (event.code === 'Enter' || event.code === 'NumpadEnter') onSubmit();
   };
 
+  console.log('calling hook')
+
   useEffect(() => {
+    console.log('calling effect')
+
     const refPointer = ref; // re-assigning so it's available on unmount
```
will show this output:

```
calling hook # first render
calling effect # first render, on mount
(5) calling hook # 5 keystrokes, re-rendering
calling hook # submitting the form and unmounting
calling effect # unmounting
```

We could use `JSON.stringify(ref)` in the dependency array, it works the same as above, but I'm worried that would have an unintended effect. What if a property changes in the ref, but the ref is the same DOM object? We would have another event listener attached. I'm not sure if that could happen to a ref, but would rather be specific about when we want to attach the event listener to. I don't see a use case where someone uses the same function to pass different ref on a render of the same component, which is what listening for ref would support.

I get that eslint gives a warning for that (maybe we should allow `[]`?) but I want to make sure we don't introduce side effects trying to placate eslint.

Reply:
I agree that the dependency array alone is not enough to ensure that `addEventListener` happens when it should, and doesn't when it shouldn't.  But if we use the dependency array _in conjunction with_ a conditional like I mentioned above, __and__ we get that conditional right, I think that will solve our problem and it's the right way to go.

Reply:
> I feel like `useEffect` requires a new way of thinking about effects.
> 
> I found [this comment](https://github.com/facebook/react/issues/14920#issuecomment-470913287) and the related discussions from React folks helpful. (also [this one](https://github.com/facebook/react/issues/14920#issuecomment-470945051) from @gaeron himself)
> 
> Old way: How do I make `useEffect` behave like `componentDidMount?` Put `[]` in the dependency array.
> 
> New way: This function _is going to run_ on every render, but I only need to make changes when _X_ is the case. What tools do I have to ensure that?
> 
> 1. Yes, we have the dependency array, and we should put everything in there that's supposed to be there.  (`listener` and `ref` in our case)
> 2. But we also can do some old-fashioned condition checking. something like:
> 
> ```js
> if (ref && !listenerAdded) {
>   refPointer?.current?.addEventListener?.('keydown', listener);
>   listenerAdded = true;
> }
> ```
> 
> This will allow the effect to run on every render, but only make changes when needed.

Thanks, I get what you are saying, it is a different way of thinking and I'm trying to make sense of it too!

I guess i'm hung up on the behavior of `addEventListener`? \@gaeron's comment is: 
>However, if later it (props) happens that they change (e.g. if a parent component passes different function depending on state), your component will handle that correctly.

I found that our component didn't handle things correctly, another event listener would be attached as the browser thinks there are new arguments to the function. We can `JSON.stringify(ref)`, then it won't update, I'm just worried about introducing a side effect with that if a ref property updates.

I think of it more as calling to an API, which often happens on mount - we are talking the browser api and will let them handle the events from there.

I'm not opposed to `[JSON.stringify(ref)]` or some sort of flag that is set, but I just wonder why we are doing that when we know the behavior of `addEventListener` and can handle things specifically on mount and unmount

Reply:
> I found that our component didn't handle things correctly, another event listener would be attached as the browser thinks there are new arguments to the function. We can JSON.stringify(ref), then it won't update, I'm just worried about introducing a side effect with that if a ref property updates.

This is why I'm saying we need both a conditional __and__ the dependency array.

> I think of it more as calling to an API, which often happens on mount

Old way: "Call the API on mount".  New way: "Get the data when we need it-- first check to see if we have the data, and if we don't, then fetch it."

> I just wonder why we are doing that when we know the behavior of addEventListener and can handle things specifically on mount and unmount

Because in the new way of thinking, `useEffect` forces us to [handle values over time](https://github.com/facebook/react/issues/14920#issuecomment-470913287), rather than attaching effects to points in the lifecycle.

> If your component doesn't handle updates to a prop, it is usually buggy. The design of useEffect forces you to confront it. 

I think this applies here as well, even though it's not strictly a prop we're dealing with.

Reply:
> This is why I'm saying we need both a conditional **and** the dependency array.
> 
> >

Where is the conditional defined in your example? In the component?

Reply:
> We can JSON.stringify(ref), then it won't update, I'm just worried about introducing a side effect with that if a ref property updates.

hmm, could also try `JSON.stringify(ref?.current)`

Reply:
> Where is the conditional defined in your example? In the component?

Inside the `useEffect`!  The conditional would wrap the `addEventListener` call and decide if it should happen or not.

Reply:
Example from Foreman of using a conditional inside useEffect here: https://github.com/theforeman/foreman/blob/812eb536e6047c88639d08b361715781e32ff3ba/webpack/assets/javascripts/react_app/components/common/TypeAheadSelect/index.js#L28

Reply:
> Example from Foreman of using a conditional inside useEffect here:

That's coming from redux right? 

> Inside the `useEffect`! The conditional would wrap the `addEventListener` call and decide if it should happen or not.

I'm wondering how the state is going to persist each time the hook is called, once it's set to false. I'm guessing we could use a hook, but then would have to pass the getter and setter. And then we are setting to something to `true` on each mount and flipping to false on first effect run, which makes me go back to asking why not just run the effect on mount.

> Old way: "Call the API on mount". New way: "Get the data when we need it-- first check to see if we have the data, and if we don't, then fetch it."

What is the equivalent here? Checking for the specific event listener from a ref does not prove to be easy, I tried that as well, that would be a really nice solution here and I would be all for it.

I get what is being said on the issue, but a lot of it breaks down for this specific use case. It seems like there is still a lot of gray area, further down it's said `When integrating with imperative things like jQuery plugins or raw DOM APIs, some nastiness may be expected. `. 

I'm not arguing against any of the best practices, but seeing as any time the effect callback runs a new duplicate event will be added, I would really like to avoid that.

Maybe you could provide a working example? I'm trying to follow along with what you are suggesting, but can't get past the theoretical.

Reply:
interesting :thinking: I would imagine it to be bit easier to add simple event listener xD
I'm not gonna judge what's better, cuz I don't know. seeing so much additional code, I'm ok to go with the prior solution. We should not rerender the input in theory, right?
But up to you guys :) I'll merge whatever you decide :man_shrugging: xD

Reply:
@jeremylenz @ezr-ondrej the good news is I was able to add both `ref` and `listener` by using `useCallback` (basically a way to memoize a function) on the listener. I tried this yesterday, but I didn't realize `useCallback` takes a dependency array itself and that can be blank, making sure it doesn't re-create the function.

This means that the function stays the same, so we can trust the browser to not keep adding event listeners since the arguments to `addEventListener` stays the same.

If you want to poke around in the storybook chrome supports this function in devtools: `getEventListeners($('input.pf-c-form-control')[0])`

The caveat is that we now have passed the eslint buck a bit, this warning is now showing:

```
/home/vagrant/foreman/webpack/assets/javascripts/react_app/common/hooks/useSubmitOnEnter.js
  11:6  warning  React Hook useCallback has a missing dependency: 'onSubmit'. Either include it or remove the dependency array. If 'onSubmit' changes too often, find the parent c
omponent that defines it and wrap that definition in useCallback  react-hooks/exhaustive-deps
```

We could do what they are suggesting and require the passed in `onSubmit` function to be wrapped in `useCallback`, but that requires everyone using this hook to do so. I vote we ignore this warning since we now have the dependency array in `useEffect` and the specific behavior of `addEventListener` requires the function to be the same. (I can disable and comment if we go this route)

Thoughts?

Reply:
Hmm, I think using `useCallback` here avoids breaking the Hooks rules here, but then just pushes the problem up one level.  Also, without adding the conditional you still have the theoretical problem of adding the event listener too many times if `ref` happens to change a property.  I don't really see any advantage to this over the ref + conditional approach.  But let me know what I'm missing.. :thinking:

Reply:
> Also, without adding the conditional you still have the theoretical problem of adding the event listener too many times if `ref` happens to change a property. I don't really see any advantage to this over the ref + conditional approach. But let me know what I'm missing..

The big difference is the effect can run as much as it wants. As long as `ref.current` doesn't change, we just keep calling the function on the browser with the same arguments, and the browser won't create duplicate event event listeners since the function is now memoized, so it's all the same from the browser's point of view. So ref can update as much as it wants and we will handle it (since it's not using `JSON.stringify`, it's probably updating each render).

I know we still have a warning there, but I feel this fits more with the useEffect philosophy of letting the lifecycle happening and handling in the effect.

> I don't really see any advantage to this over the ref + conditional approach. But let me know what I'm missing..

Is using a ref to handle state the best approach here? If we are going to manage the state ourselves I would rather push it out to a hook or redux, but that is getting into an overly-complex area IMO

I think given the behavior of `addEventListener`, we have to pick a poison :skull: either manage lifecycle state ourselves or look the other way at a eslint warning since we know we want the function to be memoized.

Reply:
You're right, I realized that `useRef` isn't necessary to keep track of the state here.  I switched it to `useState` and it still works fine:

```js
export const useSubmitOnEnter = (ref, onSubmit) => {
  const [listenerAdded, setListenerAdded] = useState(false);

  useEffect(() => {
    const listener = event => {
      if (event.code === 'Enter' || event.code === 'NumpadEnter') onSubmit();
    };
    const refPointer = ref; // re-assigning so it's available on unmount
    if (listenerAdded === false) {
      /* eslint-disable-next-line no-unused-expressions */
      refPointer?.current?.addEventListener?.('keydown', listener);
      setListenerAdded(true);
    }

    return () =>
      refPointer?.current?.removeEventListener?.('keydown', listener);
  }, [ref, listenerAdded, onSubmit]);
};
```

After removing that objection, are there any remaining issues with this approach?  (The listener function doesn't need to be a dependency because it's defined inside the effect, so that problem is addressed too.)

Reply:
@jeremylenz ok, that works, I updated here.

For some reason I always assumed a hook inside the custom hook would always go back to the default each time the hook ran, so we would have pass one in. But I guess it works just like a hook inside a component. Good to know!

The test is now failing which I'm a bit puzzled by, but looking into that.

Thanks for discussing!

Reply:
@jeremylenz I would up having to update the test for the hook, using a small test component and testing the behavior. I think it's a bit more "natural" way of testing. The hook is also tested indirectly in the `inlineEdit` tests.

It's in a separate commit if you want to play around with the previous test and see why it's failing. I think something is going on with object reference or scoping, but it was also mocking out a ref object so was probably error-prone to begin with. I couldn't get it to work with this new method.

Let me know if there are any objections!